rgw: add s3control apis for account-wide PublicAccessBlock

cbodley opened the pull request:
we already support bucket-level PublicAccessBlock configuration through s3 PutBucketPublicAccessBlock. add support for s3control PutPublicAccessBlock to set configuration for a whole account. when both bucket and account configurations are present, take their union

rgw_enable_apis supports a new "s3control" entry that registers a RGWRESTMgr for the path "/v20180820/configuration/publicAccessBlock". because this path is hosted at the same endpoint as the s3 namespace, this would conflict with buckets named "v20180820". ~~for that reason, "s3control" is not enabled by default in rgw_enable_apis~~ presence of the request header `x-amz-account-id` is used to disambiguate

https://docs.aws.amazon.com/AmazonS3/latest/userguide/configuring-block-public-access-account.html

Fixes: https://tracker.ceph.com/issues/71265

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

cbodley committed:
rgw/s3: rename req_state::bucket_access_conf to public_access_block

the name reflects a PublicAccessBlock configuration that may either come
from the bucket (s3 PutPublicAccessBlock) or the account (s3control)

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
rgw: PublicAccessBlockConfiguration as raw struct

we really don't need encapsulation for this

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
rgw: req_state::public_access_block is not optional

a default-constructed PublicAccessBlockConfiguration (with all bool
members set to false) makes for a perfectly good empty state

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
rgw: add helper for bucket + account PublicAccessBlock config

get_public_access_conf() takes an optional account, and checks
RGW_ATTR_PUBLIC_ACCESS on that in addition to the bucket. if both attrs
are found, return the union of their configurations

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
rgw/iam: add s3:Get/PutAccountPublicAccessBlock actions

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
rgw: add s3control apis for PublicAccessBlock

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
rgw: enforce BlockPublicAcls on CreateBucket

this check is unique to account-level PublicAccessBlock configuration

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
rgw/main: clean up use of rgw_enable_apis

replace `get_str_vec()` with `ceph::split()`

replace `map<string, bool>` with `set<string>`

replace `count(key) > 0` with `contains(key)`

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
rgw/rest: add RGWRESTMgr_S3Website

move some s3website stuff out of RGWRESTMgr_S3::get_handler() into its
own nested RGWRESTMgr

Signed-off-by: Casey Bodley <cbodley@redhat.com>

adamemerson commented:
make check arm64

cbodley committed:
rgw: add s3control by default to rgw_enable_apis

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
rgw/rest: use ceph::split() for rgw_enable_apis

avoid allocating a list of strings to parse the comma-separated
rgw_enable_apis configuration

the range returned by ceph::split() has no size() function, so change
the calculation to not require it - `size() - distance(begin(), pos)`
is the same thing as `distance(pos, end())`

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
rgw/s3: apply vhost logic to s3 only

the vhost-style transformations ran in RGWREST::preprocess() before we
even route the request, so applied to every REST API in radosgw

vhost-style requests are specific to the S3 API, so they should only
apply after being routed to RGWRESTMgr_S3

extract the vhost logic from RGWREST::proprocess() into
rgw_rest_transform_s3_vhost_style(), and call that only from
RGWRESTMgr_S3::get_resource_mgr_as_default()

url-decoding of request_uri into decoded_uri is now duplicated in
preprocess() to apply to all requests, then again after vhost-style
transforms the request_uri

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley commented:
> this doesn't match the path /v20180820/ so gets misinterpreted as s3. i think we need to change the vhost logic so it only applies to s3 requests

added commits up to "rgw/s3: apply vhost logic to s3 only" to resolve this

clwluvw commented:
> i'm not sure why, but boto3 is adding the value of x-amz-account-id to the beginning of the host header for s3control requests:
> 
> > HTTP_HOST=rgw11111111111111111.localhost:8000

Apparently AWS serves s3-control API per account: https://docs.aws.amazon.com/general/latest/gr/s3.html#auto-endpoints-control-s3

cbodley commented:
rebased over conflicts with https://github.com/ceph/ceph/pull/64251

adamemerson commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -5813,30 +5814,73 @@ void parse_post_action(const std::string& post_body, req_state* s)
+{
+  // s3control apis all expect the request header x-amz-account-id,
+  // and s3 apis don't. use that to disambiguate between s3control
+  // and requests to s3 buckets named v20180820
```
Good.

adamemerson commented:
jenkins test make check

cbodley commented:
> I assume this needs s3-tests as well for QA.

thanks for review! will schedule qa against https://github.com/ceph/s3-tests/pull/669

cbodley committed:
qa/rgw: s3tests task configures account ids

s3tests for s3control and iam apis need to know the account id

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
qa/rgw: RGWEndpoint.url() contains dns-name if configured

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
qa/openssl-keys: enable wildcard dns with subjectAltName

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
qa/rgw: add s3control subsuite

i would prefer to run the s3control test coverage in rgw/verify, but it
depends on rgw_dns_name configuration and support for wildcard dns which
breaks most of the other rgw/verify test cases

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley commented:
downstream testing uncovered a design bug with this feature - it looks for PublicAccessBlock configuration on the requester's account, while [PutPublicAccessBlock](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutPublicAccessBlock.html) instead refers to the bucket owner’s account:
> When Amazon S3 evaluates the PublicAccessBlock configuration for a bucket or an object, it checks the PublicAccessBlock configuration for both the bucket (or the bucket that contains the object) and the bucket owner's account.

cbodley commented:
> downstream testing uncovered a design bug with this feature - it looks for PublicAccessBlock configuration on the requester's account, while [PutPublicAccessBlock](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutPublicAccessBlock.html) instead refers to the bucket owner’s account:
> 
> > When Amazon S3 evaluates the PublicAccessBlock configuration for a bucket or an object, it checks the PublicAccessBlock configuration for both the bucket (or the bucket that contains the object) and the bucket owner's account.

updated to address this. the `get_public_access_conf()` helper was renamed to `read_public_access_conf()`, and tries to read the bucket owner's account info/xattrs

cbodley commented on src/rgw/rgw_rest_s3control.cc:
```diff
@@ -0,0 +1,339 @@
+
+int RGWDeletePublicAccessBlock_S3Control::verify_permission(optional_yield y)
+{
+  const rgw::ARN arn{"", "root", account_id, false}; // XXX
```
in https://github.com/ceph/ceph/pull/68429 i had added a `rgw::account::root_arn()` helper function for exactly this, so i've updated this to use it

cbodley committed:
qa/rgw: ignore 'keytool: command for found' errors

this 'keytool' invocation was moved from qa/tasks/s3tests_java.py to
qa/tasks/rgw.py so that it would also cover the java checksum tests

but that means it runs for any rgw job with https enabled, even if it
doesn't install or use any java stuff. the 'keytool' command itself
comes from jdk packages which aren't installed by default

ignore errors from this command so that subsuites can use https without
installing java

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley commented:
https://jenkins.ceph.com/job/ceph-pull-requests-arm64/94200/
> The following tests FAILED:
	273 - unittest_posix_bucket_cache (Timeout)

cbodley commented:
jenkins test make check arm64

cbodley commented:
passed qa in https://pulpito.ceph.com/cbodley-2026-05-21_18:30:35-rgw-wip-qa-71265-distro-default-trial/

cbodley committed:
rgw/s3control: skip account id check for admin users

allow access to admin users that don't belong to the requested account.
this is also necessary for multisite, where requests are forwarded to
the metadata master as the multisite system user instead of the original
requester

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley commented:
i've raised a separate pr https://github.com/ceph/ceph/pull/69111 for documentation changes so they can be reviewed without delaying merge here

batrick commented:
Please share regular updates on this `umbrella` branch blocker in `#ceph-upstream-releases-umbrella` on Slack.
